### ISSUE TITLE
Add ignore: ["consecutive-duplicates"] option to declaration-block-no-duplicate-properties

### ISSUE THREAD

**Issue by arkhi** [stylelint/stylelint#1021]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

An option `except: "consecutive"`

It is better explained in [csslint documentation for the rule `duplicate-properties`](https://github.com/CSSLint/csslint/wiki/Disallow-duplicate-properties).

> Which rule, if any, is this issue related to?

`declaration-block-no-duplicate-properties`

> What CSS is needed to reproduce this issue?

e.g.

```css
.foo {
    background-color: #fff;
    background-color: rgba(255, 255, 255, 0.5);
}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
  "rules": {
    "declaration-block-no-duplicate-properties": true
  }
}
```

> Which version of stylelint are you using?

`4.4.0`

**Comment by jeddy3** [comment]
@arkhi Thanks for the suggestion. This sounds like a good addition to me. Even though I suspect a lot of people use [tooling](https://github.com/postcss/postcss-color-rgba-fallback) for these types of duplicates, there is a valid use case here. 

Is this something you'd like to champion and contribute? We've [written up a guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md) to help new contributors get started :) Also, the [`selector-no-type`](https://github.com/stylelint/stylelint/tree/master/src/rules/selector-no-type) is great example of working with the `ignore` option and has nearly everything needed to get this done: [options validator](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-no-type/index.js#L20-L27), [conditionals](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-no-type/index.js#L51-L53), [tests](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-no-type/__tests__/index.js#L77-L103) and [docs](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-no-type/README.md#optional-options).

I think the option would be `ignore: ["consecutive-duplicates"]`, as it'll keep things aligned with a [pending issue](https://github.com/stylelint/stylelint/issues/882) for a similar option for this rule.

Can anyone else see any issues with adding this option i.e. just checking if there's a previous declaration and whether the property of that declaration matches? I'll mark this issue as "option" and "help wanted", but shout if you see a problem with it.

**Comment by ntwb** [comment]
I also have *not* enabled the `declaration-block-no-duplicate-properties` rule for this same reason.

Previously I had thought the [no-unsupported-browser-features](https://github.com/stylelint/stylelint/tree/master/src/rules/no-unsupported-browser-features) might be a canditate here as `doiuse` would check `caniuse` for my `browserlist` array of supported browsers for those browsers that did or did not include support for alpha transparency colors/background-colors or not.

Apart from my above initial thoughts I've not yet explored this fully yet either, so unsure if this would work or not work or just plain craziness :smirk:

**Comment by evilebottnawi** [comment]
@jeddy3 what do you think rename to `ignore: ["fallback-duplicates"]`, have array properties which can be fallback (in rule)?

**Comment by ntwb** [comment]
Yes, `ignore: ["fallback-duplicates"]` would be more appropriate.

For example the following CSS would warn of this, yet they are not actually `consecutive`:
```
.foo {
    background-color: #fff;
    color: #000;
    background-color: rgba(255, 255, 255, 0.5);
}
```

**Comment by evilebottnawi** [comment]
@jeddy3 @ntwb with `ignore: ["fallback-duplicates"]` i can make feature today and PR this.

**Comment by arkhi** [comment]
@jeddy3: Thanks a lot for all the links and hints! That’s really inviting to try it. :)

I just started testing postcss and stylelint today, so even if I’m interested to give it a go, I wouldn’t like to stop anybody faster to do it. So if nobody has a PR anytime soon, I’ll give it a shot.

For the sake of discussion, I would still recommend `consecutive` properties instead of the [example @ntwb was mentioning](https://github.com/stylelint/stylelint/issues/1021#issuecomment-206326940). I’m all new to stylelint, so maybe there are conventions I don’t know about which would explain this choice of not keeping them consecutive, but allowing separate duplicate properties feels prone to cause confusion (and bugs) to me.

**Comment by arkhi** [comment]
That being said, it can be two different ignore options. :)

**Comment by jeddy3** [comment]
Lets work this through as I think there might be more to this option than we originally thought.

I think we should start by clarifying the terminology we're using. I am I right in thinking that browsers *ignore* declaration they don't understand, either because the browser doesn't recognise the property, the value or the unit? "Fallbacks" are when developers knowingly use this mechanism by including similar declarations and anticipating one or more of them will be ignored by particular browsers.   

As such, I'm not sure it is viable to have an "array [of] properties which can be fallback (in rule)?" as that list will encompass most properties right? For example, any property who value is a *dimension* would likely be included (e.g. `rem` -> `px`). Likewise, any property that accepts a *color* (e.g. `rgba()` -> `hex`).

I think the end game is to have, as @ntwb suggested, an option that uses `doiuse` to work out whether a certain value or unit is supported by the targeted browsers and then anticipating a fallback being used. I think this will be a non-trivial addition and something we can add as another option at the later date.

In the meantime, I think we have two proposals that can be clearly defined and address the most obvious use cases:

- `"ignore": ["consecutive-duplicates"]`
- `"ignore": ["duplicates-with-prefixed-values"]`

A user can *opt-in* to either or both of these options on their use case. For example, `consecutive-duplicates` would work well with the `declaration-block-properties-order` rule (with a *strict* configuration). Where as `duplicates-with-prefixed-values` would work independently of any other rules.

What do you all think?

> I just started testing postcss and stylelint today, so even if I’m interested to give it a go,

We're always eager to welcome new contributors :) @evilebottnawi As such, can you give @arkhi the opportunity to get involve and can you hold off doing a PR yourself for now?

**Comment by evilebottnawi** [comment]
@jeddy3 yep :smile:

**Comment by arkhi** [comment]
I’m looking at the rule and I get the idea of what should be done. Now I need to RTFM for postcss. :)

**Comment by davidtheclark** [comment]
I think `"consecutive-duplicates"` is the best option here. Because it is simple, and I think it addresses the need while also enforcing a best practice.

@ntwb I think the CSS example you gave would be significantly less confusing to read if the fallback properties were consecutive, so I think that that's what we should enforce.

`"duplicates-with-prefixed-values"` wouldn't even allow for fallbacks, because often those fallbacks don't involve prefixes at all (e.g. `rem` & `px`).

**Comment by jeddy3** [comment]
> I think "consecutive-duplicates" is the best option here. Because it is simple, and I think it addresses the need while also enforcing a best practice.

Agreed. 

@arkhi Feel free to fire over a PR for this option. Also, please don't hesitate to push up a work-in-progress PR if you want early feedback or have any questions. Good luck :)

**Comment by ntwb** [comment]
Cool :+1: 

My reasoning above was more on _detection_ and not so much on _conventions_, I would hope most devs indeed listed these consecutively, as such my thinking was that we should be able to detect these unconventional uses. But lets go with `consecutive-duplicates`, if it turns out people are indeed _unconventional_ then iterating unconventional uses cases down the road once actually needed.

**Comment by arkhi** [comment]
I’ll gladly make a PR tomorrow or Saturday if that’s OK. I’ll surely have some questions. :)
